Reading csv file

In [3]:
import pandas as pd

df = pd.read_csv("Combined_dataset.csv")

Understanding the dataset

In [4]:
df.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,"""₹3,995.00""",INR,http://assets.myntassets.com/assets/images/837...,...,"{""1_star"":2,""2_stars"":0,""3_stars"":3,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-8376765_cae700""]",NaN,[{}],{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/913...,...,"{""1_star"":3,""2_stars"":4,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-9136281_cae700""]",NaN,"[{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,"""₹2,999.00""",INR,http://assets.myntassets.com/assets/images/176...,...,"{""1_star"":9,""2_stars"":5,""3_stars"":10,""4_stars""...",NaN,NaN,"[{""size"":""S""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,"[{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,"""₹1,675.00""",INR,http://assets.myntassets.com/assets/images/137...,...,"{""1_star"":33,""2_stars"":21,""3_stars"":69,""4_star...",NaN,NaN,"[{""size"":""Onesize""}]","[""https://videos.myntassets.com/assets/videos/...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",backpacks
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,"""₹2,899.00""",INR,http://assets.myntassets.com/assets/images/pro...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NaN,NaN,"[{},{},{},{},{},{},{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on IndusI...",bath-robe


In [5]:
df.tail()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,images,...,amount_of_stars,what_customers_said,seller_name,sizes,videos,seller_information,variations,best_offer,more_offers,category
995,13698258,KREDO,Men FORTE Silver Automatic Skeleton Analogue W...,3.5,11,11890,50.0,"""₹11,890.00""",INR,http://assets.myntassets.com/assets/images/136...,...,"{""1_star"":3,""2_stars"":0,""3_stars"":2,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""pi-13698258""]",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
996,16896006,GUESS,Women White Patterned Dial & White Straps Anal...,5.0,22,8900,45.0,"""₹8,900.00""",INR,http://assets.myntassets.com/assets/images/168...,...,"{""1_star"":0,""2_stars"":0,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]","[""rw-16896006_cae700"",""https://videos.myntasse...",NaN,NaN,{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
997,1708108,GIORDANO,Women Rose Gold-Toned Mirrored Analogue Watch ...,2.8,5,7950,50.0,"""₹7,950.00""",INR,http://assets.myntassets.com/assets/images/170...,...,"{""1_star"":2,""2_stars"":1,""3_stars"":0,""4_stars"":...",NaN,NaN,"[{""size"":""Onesize""}]",NaN,NaN,"[{},{},{},{},{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",watches
998,19280114,Fastrack,Men Black Brass Dial & Black Stainless Steel S...,4.3,17,3295,20.0,"""₹2,636.00""",INR,http://assets.myntassets.com/assets/images/192...,...,"{""1_star"":1,""2_stars"":2,""3_stars"":0,""4_stars"":...",NaN,Tru███m R███il,"[{""size"":""Onesize""}]",NaN,Sho███o. ███-10█████████enu███████████████████...,NaN,"{""applicable_on"":""Orders above Rs. 1499 (only ...","[{""offer_name"":""10% Instant Discount on IndusI...",watches
999,18601682,BOLDFIT,Purple Textured Yoga Mat,3.8,161,699,31.0,"""₹699.00""",INR,http://assets.myntassets.com/assets/images/186...,...,"{""1_star"":24,""2_stars"":8,""3_stars"":22,""4_stars...",NaN,NaN,"[{""size"":""Medium""}]",NaN,NaN,"[{},{},{},{}]",{},"[{""offer_name"":""10% Instant Discount on Citi C...",yoga-mats


In [6]:
df['best_offer'].head(15)

,best_offer
0,{}
1,{}
2,{}
3,{}
4,{}
5,"{""applicable_on"":""Orders above Rs. 1499 (only ..."
6,{}
7,{}
8,{}
9,{}


In [7]:
df['more_offers'].head(15)

,more_offers
0,"[{""offer_name"":""10% Instant Discount on Citi C..."
1,"[{""offer_name"":""10% Instant Discount on Citi C..."
2,"[{""offer_name"":""10% Instant Discount on Citi C..."
3,"[{""offer_name"":""10% Instant Discount on Citi C..."
4,"[{""offer_name"":""10% Instant Discount on IndusI..."
5,"[{""offer_name"":""10% Instant Discount on IndusI..."
6,"[{""offer_name"":""10% Instant Discount on Citi C..."
7,"[{""offer_name"":""10% Instant Discount on Citi C..."
8,"[{""offer_name"":""10% Instant Discount on Citi C..."
9,"[{""offer_name"":""10% Instant Discount on Citi C..."


In [8]:
df['breadcrumbs'].head(10)

,breadcrumbs
0,"[{""name"":""Accessories"",""url"":""https://www.mynt..."
1,"[{""name"":""Accessories"",""url"":""https://www.mynt..."
2,"[{""name"":""Accessories"",""url"":""https://www.mynt..."
3,"[{""name"":""Accessories"",""url"":""https://www.mynt..."
4,"[{""name"":""Men"",""url"":""https://www.myntra.com/m..."
5,"[{""name"":""Unisex"",""url"":""https://www.myntra.co..."
6,"[{""name"":""Unisex"",""url"":""https://www.myntra.co..."
7,"[{""name"":""Unisex"",""url"":""https://www.myntra.co..."
8,"[{""name"":""Unisex"",""url"":""https://www.myntra.co..."
9,"[{""name"":""Unisex"",""url"":""https://www.myntra.co..."


In [9]:
df.shape

(1000, 24)

In [10]:
df.dtypes

,0
product_id,int64
title,object
product_description,object
rating,float64
ratings_count,int64
initial_price,int64
discount,float64
final_price,object
currency,object
images,object


In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   product_id              1000 non-null   int64  
 1   title                   1000 non-null   object 
 2   product_description     1000 non-null   object 
 3   rating                  1000 non-null   float64
 4   ratings_count           1000 non-null   int64  
 5   initial_price           1000 non-null   int64  
 6   discount                879 non-null    float64
 7   final_price             1000 non-null   object 
 8   currency                1000 non-null   object 
 9   images                  1000 non-null   object 
 10  delivery_options        1000 non-null   object 
 11  product_details         1000 non-null   object 
 12  breadcrumbs             1000 non-null   object 
 13  product_specifications  1000 non-null   object 
 14  amount_of_stars         1000 non-null   o

In [12]:
df.describe()

,product_id,rating,ratings_count,initial_price,discount
count,1.000000e+03,1000.0000,1000.000000,1000.00000,879.000000
mean,1.713437e+07,3.6215,76.691000,2723.24100,53.503982
std,3.813766e+06,1.3744,241.114263,2408.69736,18.029201
min,5.868460e+05,0.0000,0.000000,249.00000,1.000000
25%,1.511501e+07,3.7000,7.000000,1399.00000,44.000000
50%,1.820890e+07,4.1000,17.000000,1999.00000,56.000000
75%,1.902737e+07,4.3000,58.000000,3299.00000,66.000000
max,2.274707e+07,5.0000,4441.000000,22199.00000,88.000000


In [13]:
df.isnull().sum()

,0
product_id,0
title,0
product_description,0
rating,0
ratings_count,0
initial_price,0
discount,121
final_price,0
currency,0
images,0


Here we see that following columns contain missing values:
discount(121),
what_customer_said(573),
seller_name(301),,
videos(781),
seller_information(301),
variations(562).

discount: discount is a numeric column and contains few missing values (121 only) so I will replace the missing values with the median of the column.

In [14]:
df['discount'].fillna(df['discount'].median(), inplace=True)

/tmp/ipykernel_6890/3279814626.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['discount'].fillna(df['discount'].median(), inplace=True)


seller_name & seller_information: Both contains 301 missing values and maybe useful later so I will fill missing values for these columns with NA

In [15]:
df['seller_name'].fillna("NA", inplace=True)
df['seller_information'].fillna("NA", inplace=True)

/tmp/ipykernel_6890/4016148684.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['seller_name'].fillna("NA", inplace=True)
/tmp/ipykernel_6890/4016148684.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try usin

what_customer_said & variations: both contain nearly half missing values. what_customer_said maybe an important feature for tasks like sentiment analysis hence I am keeping it. variations may not be an important feature and because it contains lots of missing values, I am removing it.

In [16]:
df['what_customers_said']=df['what_customers_said'].fillna("NA")
df.drop(columns=['variations'], inplace=True)

videos: videos has 781 missing values which is a lot so I am dropping it.

In [17]:
df.drop(columns=['videos'], inplace=True)

In [18]:
df.isnull().sum()

,0
product_id,0
title,0
product_description,0
rating,0
ratings_count,0
initial_price,0
discount,0
final_price,0
currency,0
images,0


Now I will create a new data frame with useful features. Here are the features that I am not using in the new dataset:

1.   image: it contains data in the form of url which is not ideal for ml model
2.   amount_of_stars : the information provided by this column is already provided by ratings. Hence it is a redundant column and is removed.
3.   breadcrumbs: the information provided by this column is already provided by category column. Hence it is a redundant column and is removed.It also contains urls
4.   best_offer: it mostly contains {} empty values and other rows give information already provided by discount and final_price
5. more_offers : contain information already given by discount and final_price hence removed,






In [19]:
useful_features = [
    "product_id",
    "title",
    "product_description",
    "rating",
    "ratings_count",
    "initial_price",
    "discount",
    "final_price",
    "currency",
    "delivery_options",
    "product_details",
    "product_specifications",
    "what_customers_said",
    "seller_name",
    "sizes",
    "seller_information",
    "category"
]

df_new = df[useful_features].copy()

Here are some row filtered datasets:
1.) high_rated : contains products with ratings more than 4.5. Useful for extracting highly rated products.
2.) high_discount : contains products with discount more than 50%. Useful for knowing which products have high discounts on them.

In [20]:
high_rated = df_new[df_new['rating'] > 4.5]
high_discount = df_new[df_new['discount'] > 50]

Checking duplicate rows and removing them

In [21]:
df.duplicated().sum()

np.int64(0)

No duplicate records were found

Creating a derived column : total_amount = price * quantity\
Since our dataset doesnt have a quantity column I will create a new column quantity filled with random values.\



In [22]:
import numpy as np
np.random.seed(42)
df_new['quantity']=np.random.randint(1,10,size=len(df_new))

Also, final_price datatype is object so I will convert it to float.

In [31]:
df_new["final_price"] = (df_new["final_price"].str.replace("₹", "", regex=False))
df_new["final_price"] = (df_new["final_price"].str.replace(",", "", regex=False))
df_new["final_price"] = (df_new["final_price"].str.replace('"', "", regex=False))
df_new['final_price']=pd.to_numeric(df_new['final_price'])

In [32]:
df_new['final_price'].head()

,final_price
0,3995.0
1,2899.0
2,2999.0
3,1675.0
4,2899.0


Finally, new feature total_amount

In [33]:
df_new['final_price']=df_new['quantity']*df_new['final_price']

In [34]:
df_new.head()

,product_id,title,product_description,rating,ratings_count,initial_price,discount,final_price,currency,delivery_options,product_details,product_specifications,what_customers_said,seller_name,sizes,seller_information,category,quantity
0,8376765,Lino Perros,Women Navy Blue Solid Backpack,3.8,15,3995,58.0,27965.0,INR,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy Blue solid backpackNon-Pa...","[{""specification_name"":""Add-Ons"",""specificatio...",NA,NA,"[{""size"":""Onesize""}]",NA,backpacks,7
1,9136281,Tommy Hilfiger,Unisex Navy Blue Striped Backpack,4.5,67,2899,35.0,11596.0,INR,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy Blue backpackNon-Padded h...","[{""specification_name"":""Add-Ons"",""specificatio...",NA,NA,"[{""size"":""Onesize""}]",NA,backpacks,4
2,17633752,Lavie,Aries Women Pink Mini Backpack,4.4,226,2999,65.0,23992.0,INR,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Pink solid backpacks Non-padd...","[{""specification_name"":""Add-Ons"",""specificatio...",NA,NA,"[{""size"":""S""}]",NA,backpacks,8
3,1376949,F Gear,Unisex Navy & Grey Printed Burner Backpack,4.4,1052,1675,52.0,8375.0,INR,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Navy and grey printed backpack...","[{""specification_name"":""Compartment Closure"",""...",NA,NA,"[{""size"":""Onesize""}]",NA,backpacks,5
4,13939916,MYTRIDENT,Men Blue Solid Bath Robe,4.7,12,2899,17.0,20293.0,INR,"[""100% Original Products"",""Pay on delivery mig...","{""description"":""Blue solid bath robe, has a sh...","[{""specification_name"":""Body or Garment Size"",...",NA,NA,"[{""size"":""M"",""value"":""45.0in"",""value_name"":""Fr...",NA,bath-robe,7


In [36]:
df_new['currency'].unique()

array(['INR'], dtype=object)

Here we can see that currency column has only one value 'INR' hence i am removing this column as it doesnt help us train model as it has a single value only

In [39]:
df_new.drop(columns=['currency'], inplace=True)

Finally we can save our data frame as a csv file

In [40]:
df_new.to_csv("cleaned_dataset.csv", index=False)